Provide kubernetes definition diffs in check_mode

willthames opened the pull request:
##### SUMMARY

Move dict_merge from azure_rm_resource module to
module_utils.common.dict_transformations and add tests.

Use dict_merge to provide a fairly realistic, reliable
diff output when k8s-based modules are run in check_mode.

Rename unit tests so that they actually run and reflect
the module_utils they're based on.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
module_utils/k8s/raw

##### ANSIBLE VERSION
```
ansible 2.7.0.dev0 (detached HEAD ref: refs/) last updated 2018/06/13 16:56:12 (GMT +1000)
  config file = None
  configured module search path = [u'/Users/will/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/will/src/ansible/lib/ansible
  executable location = /Users/will/src/ansible/bin/ansible
  python version = 2.7.14 (default, Mar  9 2018, 23:57:12) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)]
```


##### ADDITIONAL INFORMATION

fabianvf commented on lib/ansible/module_utils/k8s/raw.py:
```diff
@@ -133,7 +134,9 @@ def perform_action(self, resource, definition):
             if not existing:
-                if not self.check_mode:
+                if self.check_mode:
+                    result['result'] = k8s_obj.to_dict()
```
Is k8s_obj defined here?

fabianvf commented on lib/ansible/module_utils/k8s/raw.py:
```diff
@@ -143,7 +146,6 @@ def perform_action(self, resource, definition):
                         self.warn("{0} was not found, but creating it returned a 409 Conflict error. This can happen \
                                   if the resource you are creating does not directly create a resource of the same kind.".format(name))
                         return result
-                    result['result'] = k8s_obj.to_dict()
```
It looks like this is only added in check mode now

willthames replied:
You're quite right - not quite sure how I stuffed that up, but glad you spotted it.

Fixing this properly also fixes your other comment.

fabianvf reviewed (changes requested):
Had a few questions but like the change in general

fabianvf reviewed (approved):
:+1:

fabianvf commented:
@maxamillion I don't think I'm permissioned enough to approve this PR.

fabianvf commented:
shipit

maxamillion commented:
shipit

willthames commented:
bot_status

maxamillion commented:
rebuild_merge

jborean93 commented on lib/ansible/modules/cloud/azure/azure_rm_resource.py:
```diff
@@ -114,6 +114,7 @@
 from ansible.module_utils.azure_rm_common import AzureRMModuleBase
 from ansible.module_utils.azure_rm_common_rest import GenericRestClient
+from ansible.module_utils.common.dict_transformations import dict_merge
 from copy import deepcopy
```
Doesn't look this this is needed anymore, quick check shows that it was only used in `dict_merge`

willthames replied:
Fixed

willthames committed:
Provide kubernetes definition diffs in check_mode

Move dict_merge from azure_rm_resource module to
module_utils.common.dict_transformations and add tests.

Use dict_merge to provide a fairly realistic, reliable
diff output when k8s-based modules are run in check_mode.

Rename unit tests so that they actually run and reflect
the module_utils they're based on.

maxamillion commented:
rebuild_merge

jborean93 commented:
Errors are unrelated to the change, merging in.
